Adding UGER Support -q {long|short} and -P {projectId}

cowmoo opened the pull request:
Hi, 

I was wondering if I could submit a patch to Cromwell to support Broad's internal queue UGER, 

Currently, the GridEngine job submission backend (qsub) for Cromwell doesn't take a queue argument (-q) nor a project id (-P); both of these arguments are necessary for my group to be able to submit jobs to UGER queue successfully. 

I realize that I'm submitting this patch under 0.19_hotfix and not dev branch; and also not sure how the Cromwell team wants to organize the application.conf file to accommodate UGER arguments. 

So putting this forth initial pull request to get feedback/instructions to add this patch, 

Thanks,
Paul

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala:
```diff
@@ -25,10 +27,25 @@ import scala.sys.process._
+     } 
+    } 
+   } 
+}
```
There's existing functionality in the Cromwell-derived [Lenthall](https://github.com/broadinstitute/lenthall/blob/develop/src/main/scala/lenthall/config/ScalaConfig.scala) library to get `Option[String]` config values.

mcovarr replied:
This isn't used any more and could be deleted.

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala:
```diff
@@ -25,10 +27,25 @@ import scala.sys.process._
   object InfoKeys {
     val JobNumber = "SGE_JOB_NUMBER"
+    val QueueName = ConfigFactory.load.getConfig("backend").getConfig("sge").getOptionalString("queue")
+    val ProjectName = ConfigFactory.load.getConfig("backend").getConfig("sge").getOptionalString("project")
```
You should be able to do each of these in one shot 

```scala
val config = ConfigFactory.load
val QueueName = config.getStringOption("backend.sge.queue")
val ProjectName = config.getStringOption("backend.sge.project")
```

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala:
```diff
@@ -161,7 +178,7 @@ case class SgeBackend(actorSystem: ActorSystem) extends Backend with SharedFileS
     val logger = jobLogger(jobDescriptor)
     val sgeJobName = s"cromwell_${jobDescriptor.workflowDescriptor.shortId}_${jobDescriptor.call.unqualifiedName}"
-    val argv = Seq("qsub", "-terse", "-N", sgeJobName, "-V", "-b", "n", "-wd", jobDescriptor.callRootPath.toAbsolutePath, "-o", jobDescriptor.stdout.getFileName, "-e", jobDescriptor.stderr.getFileName, jobDescriptor.script.toAbsolutePath).map(_.toString)
+    val argv = (Seq("qsub","-terse", "-N", sgeJobName, "-V", "-b", "n", "-wd", jobDescriptor.callRootPath.toAbsolutePath) ++ InfoKeys.QueueName.map(Seq("-q",  _)).getOrElse(Seq()) ++ InfoKeys.ProjectName.map(Seq("-P", _)).getOrElse(Seq()) ++ Seq("-o", jobDescriptor.stdout.getFileName, "-e", jobDescriptor.stderr.getFileName, jobDescriptor.script.toAbsolutePath)).map(_.toString)
```
IMHO the original line was already too long, how about pulling out the new optional parameters to locals?  Also `Seq.empty` would be preferred for an empty Seq here:

```scala
val queueParam = InfoKeys.QueueName.map(Seq("-q",  _)).getOrElse(Seq.empty[String])
...
val argv = ... ++ queueParam ++ ...
```

mcovarr commented:
:+1: with comments addressed

mcovarr commented:
Thanks for the contribution!  Could you please add documentation for these new parameters to the README.md as well?

scottfrazer commented:
👍 besides @mcovarr's comments.  Very useful :)

cowmoo commented:
Thanks mcovarr and scottfrazer for your helpful suggestions/comments. 

I'm going to resubmit the patch with the suggested changes; and go from there, 

Thanks,
Paul

cowmoo commented:
Hi, I updated this patch with suggestions made by @mcovarr and also updated README.md documenting the new SGE backend optional parameters. Let me know what you guys think!

mcovarr commented:
With the unused code deleted and squashing the commits this looks good to merge.  :+1:

cowmoo commented:
Hello, I squashed all my other commits and removed the unused code. Let me know if there is still anything else I'm missing!

scottfrazer commented:
👍 thanks a lot, this is great
